qiskit/transpiler/passes/basis/basis_translator.py

```diff
@@ -294,17 +301,53 @@ def _replace_node(self, dag, node, instr_map):
                     node.op.params, node.op.name, target_params, target_dag
                 )
             )
-
         if node.op.params:
-            # Convert target to circ and back to assign_parameters, since
-            # DAGCircuits won't have a ParameterTable.
-            target_circuit = dag_to_circuit(target_dag)
-
-            target_circuit.assign_parameters(
-                dict(zip_longest(target_params, node.op.params)), inplace=True
-            )
+            parameter_map = dict(zip_longest(target_params, node.op.params))
+            bound_target_dag = target_dag.copy_empty_like()
+            for inner_node in target_dag.topological_op_nodes():
+                if any(isinstance(x, ParameterExpression) for x in inner_node.op.params):
+                    new_op = inner_node.op.copy()
+                    new_params = []
+                    for param in new_op.params:
+                        if not isinstance(param, ParameterExpression):
+                            new_params.append(param)
+                        else:
+                            bind_dict = {x: parameter_map[x] for x in param.parameters}
+                            if any(isinstance(x, ParameterExpression) for x in bind_dict.values()):
+                                new_value = param
+                                for x in bind_dict.items():
+                                    new_value = new_value.assign(*x)
+                            else:
+                                new_value = param.bind(bind_dict)
+                            if not new_value.parameters:
+                                if new_value.is_real():
+                                    new_value = (
+                                        int(new_value)
+                                        if new_value._symbol_expr.is_integer
+                                        else float(new_value)
+                                    )
+                                else:
+                                    new_value = complex(new_value)
+                            new_params.append(new_value)
+                    new_op.params = new_params
+                else:
+                    new_op = inner_node.op
+                bound_target_dag.apply_operation_back(new_op, inner_node.qargs, inner_node.cargs)
+            if isinstance(target_dag.global_phase, ParameterExpression):
+                old_phase = target_dag.global_phase
+                bind_dict = {x: parameter_map[x] for x in old_phase.parameters}
+                if any(isinstance(x, ParameterExpression) for x in bind_dict.values()):
+                    new_phase = old_phase
+                    for x in bind_dict.items():
+                        new_phase = new_phase.assign(*x)
 
-            bound_target_dag = circuit_to_dag(target_circuit)
+                else:
+                    new_phase = old_phase.bind(bind_dict)
+                try:
```

Comment:
Wouldn't it be more elegant to use the same check you did for the parameters above (i.e. `if not new_phase.parameters` and `new_phase.is_real`)? This would also allow to check for a complex global phase and raise an error in this case 🙂

Reply:
Sure, I don't remember why I did this, I think something was failing when I had the same handling as above. But I'll update the logic and raise if we have a complex global phase

Reply:
Updated in: https://github.com/Qiskit/qiskit/pull/10707/commits/b385b6059c25c520a475d50c9bbdc106297e5bde there was a little extra handling needed to account for rounding errors in symengine treating things as complex which weren't actually.